hugr-passes/src/dataflow/machine.rs

```diff
@@ -0,0 +1,131 @@
+use std::collections::HashMap;
+
+use hugr_core::{HugrView, Node, PortIndex, Wire};
+
+use super::{datalog::AscentProgram, AbstractValue, DFContext};
+
+/// Basic structure for performing an analysis. Usage:
+/// 1. Get a new instance via [Self::default()]
+/// 2. Zero or more [Self::propolutate_out_wires] with initial values
+/// 3. Exactly one [Self::run] to do the analysis
+/// 4. Results then available via [Self::read_out_wire]
+pub struct Machine<PV: AbstractValue, C: DFContext<PV>>(
+    AscentProgram<PV, C>,
+    Option<HashMap<Wire, PV>>,
+);
+
+/// derived-Default requires the context to be Defaultable, which is unnecessary
+impl<V: AbstractValue, C: DFContext<V>> Default for Machine<V, C> {
+    fn default() -> Self {
+        Self(Default::default(), None)
+    }
+}
+
+impl<PV: AbstractValue, C: DFContext<PV>> Machine<PV, C> {
+    /// Provide initial values for some wires.
+    /// (For example, if some properties of the Hugr's inputs are known.)
+    pub fn propolutate_out_wires(&mut self, wires: impl IntoIterator<Item = (Wire, PV)>) {
```

Comment:
This should be renamed - I guess maybe `populate` or `prepopulate` or just (`set`|`add`)(`_initial`|`_wire_value`)

Reply:
I've now got `prepopulate_out_wire` but also added `run` taking *input wires* as this is much easier to use.